### PR TITLE
Add foreign key constraint to all foreign keys

### PR DESCRIPTION
**Pull request by Ladas**
Add foreign key constraint to all foreign keys. Having referential integrity in the DB is crucial for having the data in a good shape.

### CODE REVIEW

**Comment by Ladas** [comment]
@agrare we should write some spec checking all foreign keys have the constraint, not sure how to do that yet

**Comment by carbonin** [comment]
@Ladas I'm not opposed to this, but it's quite a difference from what we were doing in ManageIQ. Why the change? Is this just to get the cascading delete on the database side rather than using AR associations for that?

**Comment by Ladas** [comment]
@carbonin Yes :-) We'll not use rails for DB operations :-) A lot of things will be different than in ManageIQ I suppose. :-)

**Comment by carbonin** [comment]
Yeah, that's fine, I agree with you, but it's always good to include the "why" behind changes you're making.

File: db/migrate/20181018085055_add_foreign_keys_to_all_models.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
+class AddForeignKeysToAllModels < ActiveRecord::Migration[5.1]
+  def change
+    # On cascade delete
```
Comment (carbonin) [inline db/migrate/20181018085055_add_foreign_keys_to_all_models.rb]: "On delete cascade" I think makes more sense.

File: db/migrate/20181018085055_add_foreign_keys_to_all_models.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+    add_foreign_key :container_groups, :container_nodes, on_delete: :cascade
+
+    add_foreign_key :sources, :tenants, on_delete: :cascade
+
+    add_foreign_key :authentications, :tenants, on_delete: :cascade
+
+    add_foreign_key :service_parameters_sets, :service_offerings, on_delete: :cascade
+
+    # On cascade nullify
```
Comment (carbonin) [inline db/migrate/20181018085055_add_foreign_keys_to_all_models.rb]: "On delete nullify" I think makes more sense.

Comment (Ladas) [reply db/migrate/20181018085055_add_foreign_keys_to_all_models.rb]: right

File: app/models/service_parameters_set.rb [hunk 3]
```diff
@@ -3,5 +3,5 @@ class ServiceParametersSet < ApplicationRecord
   belongs_to :source
   belongs_to :service_offering
 
-  has_many   :service_instances, :dependent => :nullify
```
Comment (Ladas) [inline app/models/service_parameters_set.rb]: @agrare I've changed this to on delete cascade, I think a plan can't live without offering

File: app/models/container_node.rb [hunk 4]
```diff
@@ -1,5 +1,5 @@
 class ContainerNode < ApplicationRecord
   belongs_to :tenant
   belongs_to :source
-  has_many   :container_groups, :dependent => :nullify
```
Comment (Ladas) [inline app/models/container_node.rb]: @agrare changed this to delete, I think pod can't live without a node to run on?

**Comment by Ladas** [comment]
@carbonin ok, added more description to the PR. :-)

File: app/models/service_parameters_set.rb (hunk 3 above)
Comment (agrare) [reply app/models/service_parameters_set.rb]: Possibly?  They are a separate top-level collection though so deletion should probably be handled by refresh.

**Pull request merged**